File: app/src/main/java/com/example/mohassu/Adapter/FriendAdapter.java
//package com.example.mohassu.Adapter;
//
//import android.content.Context;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.ImageView;
//import android.widget.TextView;
//
//import androidx.annotation.NonNull;
//import androidx.recyclerview.widget.RecyclerView;
//
//import com.bumptech.glide.Glide;
//import com.example.mohassu.Model.Friend;
//import com.example.mohassu.R;
//
//
//import java.util.ArrayList;
//import java.util.List;
//
//public class FriendAdapter extends RecyclerView.Adapter<FriendAdapter.FriendViewHolder> {
//    private List<Friend> friendList; // 전체 친구 목록
//    private List<Friend> filteredList; // 필터링된 친구 목록
//    private Context context;
//    private OnFriendClickListener onFriendClickListener;
//
//    public FriendAdapter(Context context, List<Friend> friendList, OnFriendClickListener listener) {
//        this.context = context;
//        this.friendList = new ArrayList<>(friendList);
//        this.filteredList = new ArrayList<>(friendList);
//        this.onFriendClickListener = listener;
//    }
//
//    @NonNull
//    @Override
//    public FriendViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
//        View view = LayoutInflater.from(context).inflate(R.layout.view_friend, parent, false);
//        return new FriendViewHolder(view);
//    }
//
//    @Override
//    public void onBindViewHolder(@NonNull FriendViewHolder holder, int position) {
//        Friend friend = filteredList.get(position);
//        holder.nicknameTextView.setText(friend.getNickname());
//
//        if (friend.getStatusMessage() != null && !friend.getStatusMessage().isEmpty())
//            holder.statusTextView.setText(friend.getStatusMessage());
//        else
//            holder.statusTextView.setText("상태 메시지가 없어요!");
//
//        if(friend.getCurrentClass() != null){
//            holder.placeTextView.setText(friend.getCurrentClass().classPlace + "에서 " + friend.getCurrentClass().classTitle + "수업 중!!!");
//            holder.timeTextView.setText(friend.getCurrentClass().startTime.hour + "시 " + friend.getCurrentClass().startTime.minute +"분 부터 " + friend.getCurrentClass().endTime.hour + "시 "+ friend.getCurrentClass().endTime.minute + "분 까지");
//        }
//        else{
//            holder.placeTextView.setText("지금은 수업 중이 아닌디요??");
//            holder.timeTextView.setText("친구 한테 연락해봐요!");
//        }
//
//        Glide.with(context)
//                .load(friend.getPhotoUrl())
//                .placeholder(R.drawable.img_logo)
//                .error(R.drawable.img_logo)
//                .into(holder.photoImageView);
//
//        holder.itemView.setOnClickListener(v -> {
//            if (onFriendClickListener != null) {
//                onFriendClickListener.onFriendClick(friend);
//            }
//        });
//    }
//
//    @Override
//    public int getItemCount() {
//        return filteredList.size();
//    }
//
//    // 검색 기능 추가
//    public void filter(String query) {
//        filteredList.clear();
//        if (query.isEmpty()) {
//            filteredList.addAll(friendList);
//        } else {
//            for (Friend friend : friendList) {
//                if (friend.getNickname().toLowerCase().contains(query.toLowerCase())) {
//                    filteredList.add(friend);
//                }
//            }
//        }
//        notifyDataSetChanged();
//    }
//
//    // 🔥 setData() 메서드 추가 (오류 수정)
//    public void setData(List<Friend> newFriendList) {
//        this.friendList.clear();
//        this.friendList.addAll(newFriendList);
//
//        this.filteredList.clear();
//        this.filteredList.addAll(newFriendList);
//
//        notifyDataSetChanged();
//    }
//
//    public static class FriendViewHolder extends RecyclerView.ViewHolder {
//        TextView nicknameTextView, statusTextView, placeTextView, timeTextView;
//        ImageView photoImageView;
//
//        public FriendViewHolder(@NonNull View itemView) {
//            super(itemView);
//            nicknameTextView = itemView.findViewById(R.id.nickname_text);
//            statusTextView = itemView.findViewById(R.id.state_text);
//            placeTextView = itemView.findViewById(R.id.state_place);
//            timeTextView = itemView.findViewById(R.id.state_time);
//            photoImageView = itemView.findViewById(R.id.profile_image2);
//        }
//    }
//
//    public interface OnFriendClickListener {
//        void onFriendClick(Friend friend);
//    }
//}

File: app/src/main/java/com/example/mohassu/Adapter/FriendAdapter.kt
package com.example.mohassu.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.mohassu.R
import com.example.mohassu.databinding.ViewFriendBinding
import com.example.mohassu.model.Friend

class FriendAdapter(
    private val context: Context,
    private val onFriendClick: (Friend) -> Unit
) : ListAdapter<Friend, FriendAdapter.FriendViewHolder>(FriendDiffCallback()) {

    private var originalList: List<Friend> = emptyList() // 전체 친구 목록
    private var filteredList: List<Friend> = emptyList() // 필터링된 친구 목록

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FriendViewHolder {
        val binding = ViewFriendBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return FriendViewHolder(binding)
    }

    override fun onBindViewHolder(holder: FriendViewHolder, position: Int) {
        val friend = getItem(position)
        holder.bind(friend)
    }

    inner class FriendViewHolder(private val binding: ViewFriendBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(friend: Friend) {
            binding.nicknameText.text = friend.nickname

            // 상태 메시지 설정
            binding.stateText.text = if (!friend.statusMessage.isNullOrEmpty()) {
                friend.statusMessage
            } else {
                "상태 메시지가 없어요!"
            }

            // 현재 수업 정보 설정
            friend.currentScheduleClass?.let { currentClass ->
                binding.statePlace.text = "${currentClass.classPlace}에서 ${currentClass.classTitle} 수업 중!!!"
                binding.stateTime.text = "${currentClass.startTime.hour}시 ${currentClass.startTime.minute}분 부터 " +
                        "${currentClass.endTime.hour}시 ${currentClass.endTime.minute}분 까지"
            } ?: run {
                binding.statePlace.text = "지금은 수업 중이 아닌디요??"
                binding.stateTime.text = "친구한테 연락해봐요!"
            }

            // 프로필 이미지 로드 (Glide 사용)
            Glide.with(context)
                .load(friend.photoUrl)
                .placeholder(R.drawable.img_logo)
                .error(R.drawable.img_logo)
                .into(binding.profileImage2)

            // 아이템 클릭 리스너 설정
            binding.root.setOnClickListener {
                onFriendClick(friend)
            }
        }
    }

    /**
     * DiffUtil을 사용하여 리스트 변경 최적화
     */
    class FriendDiffCallback : DiffUtil.ItemCallback<Friend>() {
        override fun areItemsTheSame(oldItem: Friend, newItem: Friend): Boolean {
            return oldItem.uid == newItem.uid
        }

        override fun areContentsTheSame(oldItem: Friend, newItem: Friend): Boolean {
            return oldItem == newItem
        }
    }

    /**
     * 🔥 검색 기능 추가 (최적화된 필터링)
     */
    fun filter(query: String) {
        filteredList = if (query.isEmpty()) {
            originalList
        } else {
            originalList.filter { it.nickname.contains(query, ignoreCase = true) }
        }
        submitList(filteredList)
    }

    /**
     * 🔥 setData() 추가 (전체 목록 업데이트 시 사용)
     */
    fun setData(newFriendList: List<Friend>) {
        originalList = newFriendList
        filter("") // 초기화 후 전체 리스트 표시
    }
}
